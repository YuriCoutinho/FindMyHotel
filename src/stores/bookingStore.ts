import { defineStore } from 'pinia'

export const useBookingStore = defineStore('bookingStore', {
  state: () => ({
    selectedHotel: null as any | null,
  }),
  getters: {
    getSelectedHotel: (state) => state.selectedHotel,
  },
  actions: {
    setSelectedHotel(hotel: any) {
      this.selectedHotel = hotel
    }
  }
})
