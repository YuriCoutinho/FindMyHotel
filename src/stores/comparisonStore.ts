import { defineStore } from 'pinia'

export const useComparisonStore = defineStore('comparisonStore', {
  state: () => ({
    hotelsToCompare: [] as any[]
  }),
  actions: {
    addHotel(hotel: any) {
      if (!this.hotelsToCompare.find((h) => h.id === hotel.id)) {
        this.hotelsToCompare.push(hotel)
      }
    },
    removeHotel(hotelId: string) {
      this.hotelsToCompare = this.hotelsToCompare.filter((hotel) => hotel.id !== hotelId)
    }
  }
})
