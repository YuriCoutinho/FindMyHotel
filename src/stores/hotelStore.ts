import { defineStore } from 'pinia'
import axios from 'axios'
import { hotels as initialHotels } from '@/utils/constants'

export const useHotelStore = defineStore('hotelStore', {
  state: () => ({
    selectedState: null as { name: string; abbreviation: string } | null,
    selectedCity: '' as string | null,
    cities: [] as string[],
    checkInDate: '' as string | null,
    checkOutDate: '' as string | null,
    guests: 1,
    rooms: 1,
    searchIsLoading: false as boolean,
    hotels: [] as any[],
    originalHotels: [] as any[]
  }),
  getters: {
    getSelectedState: (state) => state.selectedState,
    getSelectedCity: (state) => state.selectedCity,
    getCheckInDate: (state) => state.checkInDate,
    getCheckOutDate: (state) => state.checkOutDate,
    getGuests: (state) => state.guests,
    getRooms: (state) => state.rooms,
    getSearchIsLoading: (state) => state.searchIsLoading,
    getHotels: (state) => state.hotels,
    getCities: (state) => state.cities,
    isCitySelectionEnabled: (state) => state.cities.length > 0,
    isFormValid: (state) => {
      return (
        state.selectedState !== null &&
        state.selectedCity !== '' &&
        state.checkInDate !== '' &&
        state.checkOutDate !== '' &&
        state.guests > 0 &&
        state.rooms > 0
      )
    }
  },
  actions: {
    setState(state: { name: string; abbreviation: string } | null) {
      this.selectedState = state
      if (state) {
        this.searchCities(state.abbreviation)
      }
    },
    setCity(city: string | null) {
      this.selectedCity = city
    },
    setCheckInDate(checkInDate: string | null) {
      this.checkInDate = checkInDate
    },
    setCheckOutDate(checkOutDate: string | null) {
      this.checkOutDate = checkOutDate
    },
    setGuests(selectedGuests: number) {
      this.guests = selectedGuests
    },
    setRooms(selectedRooms: number) {
      this.rooms = selectedRooms
    },
    setSearchIsLoading(searchIsLoading: boolean) {
      this.searchIsLoading = searchIsLoading
    },
    setHotels(hotels: any[]) {
      this.hotels = hotels
    },
    sortHotelsByPriceAsc() {
      const sortedHotels = [...this.hotels].sort((a, b) => a.price - b.price)
      this.setHotels(sortedHotels)
    },
    sortHotelsByPriceDesc() {
      const sortedHotels = [...this.hotels].sort((a, b) => b.price - a.price)
      this.setHotels(sortedHotels)
    },
    sortHotelsByStarsAsc() {
      const sortedHotels = [...this.hotels].sort((a, b) => a.stars - b.stars)
      this.setHotels(sortedHotels)
    },
    sortHotelsByStarsDesc() {
      const sortedHotels = [...this.hotels].sort((a, b) => b.stars - a.stars)
      this.setHotels(sortedHotels)
    },
    resetHotels() {
      this.setHotels([...this.originalHotels])
    },
    async searchCities(stateAbbreviation: string) {
      this.selectedCity = null
      this.cities = []
      if (stateAbbreviation) {
        try {
          const response = await axios.get(
            `https://servicodados.ibge.gov.br/api/v1/localidades/estados/${stateAbbreviation}/municipios`
          )
          this.cities = response.data.map((municipality: any) => municipality.nome)
        } catch (error) {
          console.error('Error fetching municipalities:', error)
          this.cities = []
        }
      }
    },
    async searchHotels() {
      this.setSearchIsLoading(true)
      try {
        setTimeout(() => {
          this.hotels = [...initialHotels]
          this.originalHotels = [...initialHotels]
          this.setSearchIsLoading(false)
        }, 3000)
      } catch (error) {
        console.error('Error searching hotels:', error)
        this.setSearchIsLoading(false)
      }
    }
  }
})
