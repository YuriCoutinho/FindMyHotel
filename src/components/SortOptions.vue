<template>
  <div class="flex items-center justify-between w-full">
    <Select
      class="w-44 flex items-center justify-between px-3 py-2 cursor-pointer border border-gray-300 rounded-lg"
      inputClass="w-38 text-gray-700 font-medium focus:outline-none"
      panelClass="bg-white shadow-xl rounded-lg mt-1 p-3 cursor-pointer "
      optionLabel="label"
      optionValue="value"
      placeholder="Ordenar por"
      :options="options"
      v-model="selectedOption"
    />

    <Button
      @click="resetHotels"
      label="Resetar ordem"
      icon="pi pi-refresh"
      class="flex items-center gap-2 px-3 py-2 cursor-pointer border border-gray-300 rounded-lg hover:bg-gray-400 active:bg-gray-500"
    />
  </div>
</template>

<script setup lang="ts">
import { ref, watch } from 'vue'
import Select from 'primevue/select'
import Button from 'primevue/button'

import { useHotelStore } from '@/stores/hotelStore'

const hotelStore = useHotelStore()

interface Option {
  label: string
  value: string
}

const options: Option[] = [
  { label: 'Menores preços', value: 'priceAsc' },
  { label: 'Maiores preços', value: 'priceDesc' },
  { label: 'Menos estrelas', value: 'starsAsc' },
  { label: 'Mais estrelas', value: 'starsDesc' }
]

const selectedOption = ref<string>('')

watch(selectedOption, (newValue) => {
  sortHotels(newValue)
})

const sortHotels = (value: string) => {
  switch (value) {
    case 'priceAsc':
      hotelStore.sortHotelsByPriceAsc()
      break
    case 'priceDesc':
      hotelStore.sortHotelsByPriceDesc()
      break
    case 'starsAsc':
      hotelStore.sortHotelsByStarsAsc()
      break
    case 'starsDesc':
      hotelStore.sortHotelsByStarsDesc()
      break
  }
}

const resetHotels = () => {
  hotelStore.resetHotels()
  selectedOption.value = ''
}
</script>
