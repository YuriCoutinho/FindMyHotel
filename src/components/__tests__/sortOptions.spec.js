import { mount } from '@vue/test-utils'
import { describe, it, expect, beforeEach, vi } from 'vitest'
import { createPinia, setActivePinia } from 'pinia'
import SortOptions from '@/components/SortOptions.vue'
import PrimeVue from 'primevue/config'
import Select from 'primevue/select'
import Button from 'primevue/button'
import { useHotelStore } from '@/stores/hotelStore'

describe('SortOptions.vue', () => {
  let sortHotelsByPriceAscMock,
    sortHotelsByPriceDescMock,
    sortHotelsByStarsAscMock,
    sortHotelsByStarsDescMock,
    resetHotelsMock

  beforeEach(() => {
    const pinia = createPinia()
    setActivePinia(pinia)

    sortHotelsByPriceAscMock = vi.fn()
    sortHotelsByPriceDescMock = vi.fn()
    sortHotelsByStarsAscMock = vi.fn()
    sortHotelsByStarsDescMock = vi.fn()
    resetHotelsMock = vi.fn()

    const store = useHotelStore(pinia)
    store.sortHotelsByPriceAsc = sortHotelsByPriceAscMock
    store.sortHotelsByPriceDesc = sortHotelsByPriceDescMock
    store.sortHotelsByStarsAsc = sortHotelsByStarsAscMock
    store.sortHotelsByStarsDesc = sortHotelsByStarsDescMock
    store.resetHotels = resetHotelsMock
  })

  const createWrapper = () => {
    return mount(SortOptions, {
      global: {
        plugins: [PrimeVue],
        components: { Select, Button }
      }
    })
  }

  it('Should render the component', () => {
    const wrapper = createWrapper()
    expect(wrapper.exists()).toBe(true)
  })

  it('Should call the appropriate sort method when a sorting option is selected', async () => {
    const wrapper = createWrapper()
    const selectComponent = wrapper.findComponent(Select)

    await selectComponent.vm.$emit('update:modelValue', 'priceAsc')
    expect(sortHotelsByPriceAscMock).toHaveBeenCalled()

    await selectComponent.vm.$emit('update:modelValue', 'priceDesc')
    expect(sortHotelsByPriceDescMock).toHaveBeenCalled()

    await selectComponent.vm.$emit('update:modelValue', 'starsAsc')
    expect(sortHotelsByStarsAscMock).toHaveBeenCalled()

    await selectComponent.vm.$emit('update:modelValue', 'starsDesc')
    expect(sortHotelsByStarsDescMock).toHaveBeenCalled()
  })

  it('Should call resetHotels when reset button is clicked', async () => {
    const wrapper = createWrapper()
    const button = wrapper.findComponent(Button)

    await button.trigger('click')
    expect(resetHotelsMock).toHaveBeenCalled()
  })

  it('Should match the snapshot', () => {
    const wrapper = createWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
})
