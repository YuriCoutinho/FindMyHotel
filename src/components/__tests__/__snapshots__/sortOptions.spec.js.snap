// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SortOptions.vue > Should match the snapshot 1`] = `
"<div class="flex items-center justify-between w-full">
  <div class="p-select p-component p-inputwrapper w-44 flex items-center justify-between px-3 py-2 cursor-pointer border border-gray-300 rounded-lg" data-pc-name="select" data-pc-section="root" pc7=""><span class="p-select-label p-placeholder w-38 text-gray-700 font-medium focus:outline-none" tabindex="0" role="combobox" aria-label="Ordenar por" aria-haspopup="listbox" aria-expanded="false" aria-controls="undefined_list" aria-disabled="false" data-pc-section="label">Ordenar por</span>
    <!---->
    <div class="p-select-dropdown" data-pc-section="dropdown"><svg width="14" height="14" viewBox="0 0 14 14" fill="none" xmlns="http://www.w3.org/2000/svg" class="p-icon p-select-dropdown-icon" aria-hidden="true" data-pc-section="dropdownicon">
        <path d="M7.01744 10.398C6.91269 10.3985 6.8089 10.378 6.71215 10.3379C6.61541 10.2977 6.52766 10.2386 6.45405 10.1641L1.13907 4.84913C1.03306 4.69404 0.985221 4.5065 1.00399 4.31958C1.02276 4.13266 1.10693 3.95838 1.24166 3.82747C1.37639 3.69655 1.55301 3.61742 1.74039 3.60402C1.92777 3.59062 2.11386 3.64382 2.26584 3.75424L7.01744 8.47394L11.769 3.75424C11.9189 3.65709 12.097 3.61306 12.2748 3.62921C12.4527 3.64535 12.6199 3.72073 12.7498 3.84328C12.8797 3.96582 12.9647 4.12842 12.9912 4.30502C13.0177 4.48162 12.9841 4.662 12.8958 4.81724L7.58083 10.1322C7.50996 10.2125 7.42344 10.2775 7.32656 10.3232C7.22968 10.3689 7.12449 10.3944 7.01744 10.398Z" fill="currentColor"></path>
      </svg></div>
    <!---->
  </div><button class="p-button p-component flex items-center gap-2 px-3 py-2 cursor-pointer border border-gray-300 rounded-lg hover:bg-gray-400 active:bg-gray-500" type="button" aria-label="Resetar ordem" data-pc-name="button" data-p-disabled="false" data-pc-section="root" pc8=""><span class="p-button-icon p-button-icon-left pi pi-refresh" data-pc-section="icon"></span><span class="p-button-label" data-pc-section="label">Resetar ordem</span>
    <!---->
  </button>
</div>"
`;
